### PR TITLE
WIP: feat(*): Allow applications to be in maintenance mode

### PR DESCRIPTION
**Pull request by nathansamson**
Closes: #3722

Note this is very much a non working wip.

TODO:
[ ] Test the basics, I just wrote the code without even testing if it made syntactical (or semantical) sense. I will try  this later on a computer with more RAM in vagrant
[ ] Add unit tests
[ ] Add documentation for controller API
[ ] Add user documentation
[ ] Change client
[ ] Add custom error/maintenance pages

If someone from the core dev team could tell me this is more or less the right way to implement this feature, let me now

### CODE REVIEW

File: controller/api/models.py [hunk 1]
```diff
@@ -410,6 +410,19 @@ def _has_hostname(image):
         escaped_command = command.replace("'", "'\\''")
         return c.run(escaped_command)
 
+    def maintenance_on(self, user):
+        msg = "{} enables miantenance mode".format(user.username)
+        log_event(self, msg)
+        _etcd_client.write('/deis/maintenance/{}'.format(id), True)
+
+    def maintenance_on(self, user):
```
Comment (kalbasit) [inline controller/api/models.py]: I believe you meant `maintenance_off` here.

Comment (nathansamson) [inline controller/api/models.py]: yeah I noticed as well. As said, thiis was done in 20m or so.

I was more intrested in general remarks of setting an etc key, and then using that in the router is a good way forward.
Also I wanted people to know I was onto this, to not do double work

**Comment by mboersma** [comment]
> I was more intrested in general remarks

@nathansamson at first glance, I think this is valid as a general approach: set a per-app `etcd` key that deis-router uses to tailor the HTTP response based on whether maintenance mode is on or not.

File: controller/api/models.py [hunk 2]
```diff
@@ -410,6 +410,19 @@ def _has_hostname(image):
         escaped_command = command.replace("'", "'\\''")
         return c.run(escaped_command)
 
+    def maintenance_on(self, user):
+        msg = "{} enables miantenance mode".format(user.username)
+        log_event(self, msg)
+        _etcd_client.write('/deis/maintenance/{}'.format(id), True)
```
Comment (mboersma) [inline controller/api/models.py]: This should probably be `.format(self.id), True`.

**Comment by nathansamson** [comment]
Closing, as I tried (but failed) to get a dev environment up & running on my computer (8GB ram).

Even when not running a UI, and do everything from a shell (effectively using almost all ram for nodes), builder never worked properly in my test VM.

I am planning to buy a new computer in the short-term with 16GB (or more) RAM, but for now I'll close the issue...

I might work on it in a few weeks time again, but for now I have no viable option to do so...

**Comment by bacongobbler** [comment]
@nathansamson most of us deploy Deis on one of the providers. DO's a pretty cheap option for testing changes. You should be able to runthe unit tests on your computer without an issue, but running the entire platform is certainly a resource hog.

**Comment by nathansamson** [comment]
Note: I've reworked on this in https://github.com/nathansamson/deis/tree/maintenance2

That version is now working (after fixing a lot of stupid bugs).

After I add documentation + some form of tests I'll reopen

**Pull request closed without merging**